storage/postgreskv/schema/2020011601_kvimplementation_switchover.up.sql

```diff
@@ -0,0 +1,17 @@
+DROP VIEW pathdata_pretty;
+DROP FUNCTION list_directory_reverse(BYTEA, BYTEA, BYTEA, INTEGER);
+DROP FUNCTION list_directory(BYTEA, BYTEA, BYTEA, INTEGER);
+DROP FUNCTION component_increment(BYTEA, INTEGER);
+DROP FUNCTION bytea_increment(BYTEA);
+DROP FUNCTION truncate_after(BYTEA, INTEGER, INTEGER);
+DROP TYPE path_and_meta;
+CREATE TABLE IF NOT EXISTS pathdata_v2 (
+                                        fullpath BYTEA PRIMARY KEY,
+                                        metadata BYTEA NOT NULL
+);
+ALTER TABLE pathdata_v2 ALTER COLUMN fullpath SET STORAGE PLAIN;
+ALTER TABLE pathdata_v2 ALTER COLUMN metadata SET STORAGE PLAIN;
+INSERT INTO pathdata_v2 (fullpath, metadata) SELECT fullpath, metadata FROM pathdata;
+DROP TABLE pathdata;
+DROP TABLE buckets;
+ALTER TABLE pathdata_v2 RENAME TO pathdata;
```

Comment:
okay, so, this is a full downtime migration. we can't do a zero downtime migration with this because in a zero downtime migration scenarios, some old api servers may be running. 

it is what it is!

Reply:
are we going to use this opportunity to get TOAST?

Reply:
oh is that SET STORAGE PLAIN?

Reply:
Yes!

Reply:
L12-13 @jtolds

Reply:
reading the docs, it looks like SET STORAGE MAIN might be better

Reply:
`(Actually, out-of-line storage will still be performed for such columns, but only as a last resort when there is no other way to make the row small enough to fit on a page.)`
This makes me thing we would not win anything. That sounds like the default to me.

Reply:
We might want to talk about increasing those values:
TOAST_MAX_CHUNK_SIZE
TOAST_TUPLE_TARGET
TOAST_TUPLE_THRESHOLD

Its quite hard to understand, if PLAIN simply does not care of these values.

Reply:
> (Actually, out-of-line storage will still be performed for such columns, but only as a last resort when there is no other way to make the row small enough to fit on a page.)
This is true of PLAIN also.

Reply:
i think we should be set with SET STORAGE MAIN

Reply:
also iirc without recompiling postgres to run with larger row limits, i'd expect the insert to just fail if it was PLAIN. I agree with MAIN. hopefully compression lets the rows stay within the necessary size limits.

Reply:
Fine, changing it to main. What about the other values?

Reply:
Probably too late at this point, and maybe it doesn't matter, but I ought to mention that we can probably make this work without downtime using rewrite rules. Maybe a rule to make queries on `pathdata` become queries on `pathdata_v2` until the new table is renamed into place, or a rule to make SELECTs on pathdata check both locations while the new table is populating. Some combination of things like that.

Reply:
Test with MAIN and that works with even 30K

Reply:
nice!